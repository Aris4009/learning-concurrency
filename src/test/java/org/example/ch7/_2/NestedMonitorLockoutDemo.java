package org.example.ch7._2;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

import lombok.extern.slf4j.Slf4j;

/**
 * @Author: Aris
 * @ClassName: NestedMonitorLockoutDemo
 * @Description: 嵌套监视器锁死
 * @Date: 2022/3/30
 **/
@Slf4j
public class NestedMonitorLockoutDemo {

	private final BlockingQueue<String> queue = new ArrayBlockingQueue<>(10);

	private int processed;

	private int accepted;

	public synchronized void accept(String msg) throws InterruptedException {
		queue.put(msg);
		accepted++;
	}

	// bad code
	public synchronized void process() throws InterruptedException {
		// take操作一直循环，导致同步监视锁一直不会释放
		log.info(queue.take());
		processed++;
	}

	// good code
//	public void process() throws InterruptedException {
//		log.info(queue.take());
//		synchronized (this) {
//			processed++;
//		}
//	}

	public void start() {
		Thread t = new WorkThread();
		t.setName("workThread");
		t.start();
	}

	public synchronized int[] getStat() {
		return new int[] { accepted, processed };
	}

	class WorkThread extends Thread {

		@Override
		public void run() {
			while (true) {
				try {
					process();
				} catch (InterruptedException e) {
					log.error(e.getMessage(), e);
					return;
				}
			}
		}
	}

	public static void main(String[] args) {
		NestedMonitorLockoutDemo demo = new NestedMonitorLockoutDemo();
		demo.start();
		for (int i = 0; i < 100000; i++) {
			try {
				demo.accept(String.valueOf(i));
				Thread.sleep(100L);
			} catch (Exception e) {
				log.error(e.getMessage(), e);
			}
		}
	}
}
