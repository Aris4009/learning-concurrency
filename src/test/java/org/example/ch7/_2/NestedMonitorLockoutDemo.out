2022-03-30 10:25:17
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.321-b07 mixed mode):

"Attach Listener" #12 daemon prio=9 os_prio=31 tid=0x00007fa9ce08f800 nid=0x5703 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"workThread" #11 prio=5 os_prio=31 tid=0x00007fa9cc0f1800 nid=0x5503 waiting on condition [0x0000700006f9c000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000076ba601a0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.ArrayBlockingQueue.take(ArrayBlockingQueue.java:403)
	at org.example.ch7._2.NestedMonitorLockoutDemo.process(NestedMonitorLockoutDemo.java:29)
	- locked <0x000000076ba5d6c8> (a org.example.ch7._2.NestedMonitorLockoutDemo)
	at org.example.ch7._2.NestedMonitorLockoutDemo$WorkThread.run(NestedMonitorLockoutDemo.java:49)

   Locked ownable synchronizers:
	- None

"Service Thread" #10 daemon prio=9 os_prio=31 tid=0x00007fa9ce026000 nid=0x4503 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C1 CompilerThread3" #9 daemon prio=9 os_prio=31 tid=0x00007fa9cb813800 nid=0x4603 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread2" #8 daemon prio=9 os_prio=31 tid=0x00007fa9cb81a000 nid=0x4803 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread1" #7 daemon prio=9 os_prio=31 tid=0x00007fa9cd852000 nid=0x4903 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread0" #6 daemon prio=9 os_prio=31 tid=0x00007fa9cf83e000 nid=0x3e03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Monitor Ctrl-Break" #5 daemon prio=5 os_prio=31 tid=0x00007fa9d004c800 nid=0x3c03 runnable [0x0000700006887000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	- locked <0x000000076aca56a0> (a java.io.InputStreamReader)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	- locked <0x000000076aca56a0> (a java.io.InputStreamReader)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at com.intellij.rt.execution.application.AppMainV2$1.run(AppMainV2.java:49)

   Locked ownable synchronizers:
	- None

"Signal Dispatcher" #4 daemon prio=9 os_prio=31 tid=0x00007fa9cd825000 nid=0x3a03 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Finalizer" #3 daemon prio=8 os_prio=31 tid=0x00007fa9cb813000 nid=0x3103 in Object.wait() [0x000070000657b000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000076ab08ee8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	- locked <0x000000076ab08ee8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

   Locked ownable synchronizers:
	- None

"Reference Handler" #2 daemon prio=10 os_prio=31 tid=0x00007fa9cd817800 nid=0x3003 in Object.wait() [0x0000700006478000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000076ab06c00> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x000000076ab06c00> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

   Locked ownable synchronizers:
	- None

"main" #1 prio=5 os_prio=31 tid=0x00007fa9ce809000 nid=0xe03 waiting for monitor entry [0x0000700005a5a000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.example.ch7._2.NestedMonitorLockoutDemo.accept(NestedMonitorLockoutDemo.java:24)
	- waiting to lock <0x000000076ba5d6c8> (a org.example.ch7._2.NestedMonitorLockoutDemo)
	at org.example.ch7._2.NestedMonitorLockoutDemo.main(NestedMonitorLockoutDemo.java:63)

   Locked ownable synchronizers:
	- None

"VM Thread" os_prio=31 tid=0x00007fa9d000a000 nid=0x2e03 runnable 

"GC task thread#0 (ParallelGC)" os_prio=31 tid=0x00007fa9ce009000 nid=0x1c07 runnable 

"GC task thread#1 (ParallelGC)" os_prio=31 tid=0x00007fa9ce80c000 nid=0x2103 runnable 

"GC task thread#2 (ParallelGC)" os_prio=31 tid=0x00007fa9cc80e800 nid=0x1f03 runnable 

"GC task thread#3 (ParallelGC)" os_prio=31 tid=0x00007fa9cd80b000 nid=0x2a03 runnable 

"GC task thread#4 (ParallelGC)" os_prio=31 tid=0x00007fa9cd80b800 nid=0x2b03 runnable 

"GC task thread#5 (ParallelGC)" os_prio=31 tid=0x00007fa9cd80c000 nid=0x5303 runnable 

"GC task thread#6 (ParallelGC)" os_prio=31 tid=0x00007fa9cd80c800 nid=0x5103 runnable 

"GC task thread#7 (ParallelGC)" os_prio=31 tid=0x00007fa9cc80f000 nid=0x2d03 runnable 

"VM Periodic Task Thread" os_prio=31 tid=0x00007fa9ce886000 nid=0x4303 waiting on condition 

JNI global references: 15

